### PR TITLE
Implement meeting closing

### PR DESCRIPTION
**Pull request by phgross**
- Added automatic generation of excerpts for every scheduled proposal.
- Copy back the excerpt to the original proposal.
- Create Proposalhistory entry when decide.
- Add excerpt box to the ProposalOverview.
- Update proposal states during closing a meeting.
- Show decision in ProposalOverview.

@deiferni please have a look ...

### CODE REVIEW

File: docs/HISTORY.txt [hunk 1]
```diff
@@ -5,6 +5,23 @@ Changelog
 4.3.1 (unreleased)
 ------------------
 
+- Implement meeting closing:
+
+  - Added automatic generation of excerpts for every scheduled proposal.
+  - Copy back the excerpt to the original proposal.
+  - Create Proposalhistory entry when decide.
```
Comment (deiferni) [inline docs/HISTORY.txt]: when decide**d**

File: opengever/meeting/command.py [hunk 2]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+    def __init__(self, meeting):
+        self.meeting = meeting
+
+    def execute(self):
+        for agenda_item in self.meeting.agenda_items:
+            if agenda_item.has_proposal():
+                self.generate_excerpt(agenda_item)
+
+    def generate_excerpt(self, agenda_item):
+        proposal_obj = agenda_item.proposal.submitted_oguid.resolve_object()
```
Comment (deiferni) [inline opengever/meeting/command.py]: could you add a `resolve_submitted_proposal` method?

File: opengever/meeting/command.py [hunk 3]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+    def execute(self):
+        for agenda_item in self.meeting.agenda_items:
+            if not agenda_item.has_proposal():
+                continue
+
+            self.decide_proposals(agenda_item.proposal)
+
+    def decide_proposals(self, proposal):
+        response = self.copy_document(proposal)
+        self.add_database_entry(proposal, response)
```
Comment (deiferni) [inline opengever/meeting/command.py]: Please extract `response['intid']` here and only pass the `intid`. I think it is not the `add_database_entry` method's responsibility to know about the previous response.

File: opengever/meeting/command.py [hunk 4]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+        session.add(excerpt)
+
+        proposal.excerpt_document = excerpt
+        session.add(proposalhistory.ProposalDecided(proposal=proposal))
+
+    def update_state(self, proposal):
+        proposal.execute_transition('scheduled-decided')
+
+    def copy_document(self, proposal):
+        document = proposal.submitted_excerpt_document.oguid.resolve_object()
```
Comment (deiferni) [inline opengever/meeting/command.py]: plz also add a resolve method

File: opengever/meeting/command.py [hunk 5]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+class CloseMeeting(object):
+
+    def __init__(self, meeting):
+        self.meeting = meeting
+
+    def execute(self):
+        GenerateExcerptsCommand(self.meeting).execute()
+        DecideProposalsCommand(self.meeting).execute()
+
+        msg = _(u'msg_meeting_successfully_closed',
```
Comment (deiferni) [inline opengever/meeting/command.py]: Please don't automatically show the message when the command is executed. This should be a separate step and called manually by the controller (or transition in our case). I'd suggest creating a `show_message` method.

File: opengever/meeting/command.py [hunk 6]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+                default=u'The meeting ${title} has been successfully closed, '
+                'the excerpts has been generated and send back to the initial'
```
Comment (deiferni) [inline opengever/meeting/command.py]: the excerpts have been generated and sent back to the initial...

File: opengever/meeting/command.py [hunk 7]
```diff
@@ -324,3 +363,78 @@ def __init__(self, source, target_admin_unit_id, target_path):
...
+    def update_state(self, proposal):
+        proposal.execute_transition('scheduled-decided')
+
+    def copy_document(self, proposal):
+        document = proposal.submitted_excerpt_document.oguid.resolve_object()
+        return OgCopyCommand(
+            document, proposal.admin_unit_id, proposal.physical_path).execute()
+
+
+class CloseMeeting(object):
```
Comment (deiferni) [inline opengever/meeting/command.py]: rename to `CloseMeetingCommand`

File: opengever/meeting/model/meeting.py [hunk 8]
```diff
@@ -131,6 +144,17 @@ def get_protocol_filename(self):
             translate(_("Protocol"), context=getRequest()),
             normalizer.normalize(self.get_title()))
 
+    def get_excerpt_title(self):
+        return u"{}-{}".format(
+            translate(_("Protocol Excerpt"), context=getRequest()),
+            self.get_title())
+
+    def get_excerpt_filename(self):
+        normalizer = getUtility(IIDNormalizer)
```
Comment (deiferni) [inline opengever/meeting/model/meeting.py]: This is the 3rd similar method maybe it's time to extract all but `translate(_("[NAME]")`

File: opengever/meeting/model/proposal.py [hunk 9]
```diff
@@ -119,6 +133,10 @@ def get_searchable_text(self):
         searchable = filter(None, [self.title, self.initial_position])
         return ' '.join([term.encode('utf-8') for term in searchable])
 
+    def get_decision(self):
+        if self.agenda_item:
+            return self.agenda_item[0].decision
```
Comment (deiferni) [inline opengever/meeting/model/proposal.py]: this should be one-to-one, not a list. might be incorrect in the current version. Please also add an explicit `return None`.

File: opengever/meeting/proposal.py [hunk 10]
```diff
@@ -256,6 +259,10 @@ def get_documents(self):
 
         return [document.getObject() for document in documents]
 
+    def get_excerpt(self):
+        model = self.load_model()
+        return model.submitted_excerpt_document.oguid.resolve_object()
```
Comment (deiferni) [inline opengever/meeting/proposal.py]: add resolve method

File: opengever/meeting/proposal.py [hunk 11]
```diff
@@ -291,6 +298,12 @@ def get_documents(self):
         documents.sort(lambda a, b: cmp(b.modified(), a.modified()))
         return documents
 
+    def get_excerpt(self):
+        model = self.load_model()
+        document = model.excerpt_document
+        if document:
+            return model.excerpt_document.oguid.resolve_object()
```
Comment (deiferni) [inline opengever/meeting/proposal.py]: add/use resolve method

File: opengever/meeting/protocol.py [hunk 12]
```diff
@@ -167,3 +167,44 @@ def add_protocol_type(self):
                 _(u'protocol', default=u'Protocol'),
                 context=getRequest())
         }
+
+
+class ExcerptProtocolData(object):
```
Comment (deiferni) [inline opengever/meeting/protocol.py]: ProtocolExcerptData?

Comment (deiferni) [inline opengever/meeting/protocol.py]: why did you copy this class? i think you could inherit from `PreProtocolData` and accomplish the same by overriding some methods. The generated ata would contain more information (i.e. participants) but unwanted data can always be omitted in the templates.

**Comment by lukasgraf** [comment]
:+1:

File: opengever/meeting/tests/test_close.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+        self.meeting = create(Builder('meeting')
+                              .having(committee=self.committee.load_model(),
+                                      start=datetime(2013, 1, 1),
+                                      location='There')
+                              .proposals([self.proposal_a, self.proposal_b]))
+        self.meeting.execute_transition('pending-held')
+
+        # set correct public url, used for generated meeting urls
+        from opengever.ogds.base.utils import get_current_admin_unit
```
Comment (deiferni) [inline opengever/meeting/tests/test_close.py]: you can move this import into module scope

File: opengever/testing/builders/dx.py [hunk 14]
```diff
@@ -239,6 +239,7 @@ def __init__(self, session):
         super(ProposalBuilder, self).__init__(session)
         self.arguments = {'title': 'Fooo'}
         self.model_arguments = None
+        self._submitted = False
```
Comment (deiferni) [inline opengever/testing/builders/dx.py]: awesome :sparkles:! could you also adapt existing tests where the transition is executed manually?

File: opengever/testing/builders/sql.py [hunk 15]
```diff
@@ -170,6 +171,27 @@ class MeetingBuilder(SqlObjectBuilder):
 
     mapped_class = Meeting
 
+    def __init__(self, session):
+        super(MeetingBuilder, self).__init__(session)
+        self._proposals = []
+
+    def proposals(self, proposals):
```
Comment (deiferni) [inline opengever/testing/builders/sql.py]: rename to `scheduled_proposals`?

**Comment by deiferni** [comment]
looks good! some comments though :wink:

File: opengever/meeting/protocol.py (hunk 12 above)
Comment (phgross) [inline opengever/meeting/protocol.py]: :+1:

File: opengever/testing/builders/dx.py (hunk 14 above)
Comment (phgross) [inline opengever/testing/builders/dx.py]: Done

**Comment by phgross** [comment]
@deiferni updated ...

**Comment by deiferni** [comment]
thank you :+1:

**Pull request merged**